### ISSUE TITLE
Tests fail to build: error: use of undeclared identifier 'isascii', etc

### ISSUE THREAD

**Issue by yurivict** [esa/pagmo2#273]
```
In file included from /usr/ports/devel/pagmo2/work/pagmo2-2.10-161-g0a589425/tests/fork_island.cpp:30:
In file included from /usr/local/include/boost/test/included/unit_test.hpp:18:
In file included from /usr/local/include/boost/test/impl/compiler_log_formatter.ipp:19:
In file included from /usr/local/include/boost/test/output/compiler_log_formatter.hpp:17:
In file included from /usr/local/include/boost/test/unit_test_log_formatter.hpp:26:
In file included from /usr/include/c++/v1/iostream:38:
In file included from /usr/include/c++/v1/ios:216:
/usr/include/c++/v1/__locale:597:16: error: use of undeclared identifier 'isascii'
        return isascii(__c) ? (__tab_[static_cast<int>(__c)] & __m) !=0 : false;
               ^
/usr/include/c++/v1/__locale:604:22: error: use of undeclared identifier 'isascii'
            *__vec = isascii(*__low) ? __tab_[static_cast<int>(*__low)] : 0;
                     ^
/usr/include/c++/v1/__locale:612:17: error: use of undeclared identifier 'isascii'
            if (isascii(*__low) && (__tab_[static_cast<int>(*__low)] & __m))
                ^
/usr/include/c++/v1/__locale:621:19: error: use of undeclared identifier 'isascii'
            if (!(isascii(*__low) && (__tab_[static_cast<int>(*__low)] & __m)))
                  ^
In file included from /usr/ports/devel/pagmo2/work/pagmo2-2.10-161-g0a589425/tests/fork_island.cpp:30:
In file included from /usr/local/include/boost/test/included/unit_test.hpp:20:
/usr/local/include/boost/test/impl/debug.ipp:425:25: error: no type named 'mkstemp' in the global namespace
    fd_holder cmd_fd( ::mkstemp( cmd_file_name ) );
                      ~~^
/usr/local/include/boost/test/impl/debug.ipp:712:25: error: use of undeclared identifier 'SIGTRAP'
    ::kill( ::getpid(), SIGTRAP );
                        ^
/usr/local/include/boost/test/impl/debug.ipp:893:36: error: no type named 'mkstemp' in the global namespace
    fd_holder init_done_lock_fd( ::mkstemp( init_done_lock_fn ) );
                                 ~~^
In file included from /usr/ports/devel/pagmo2/work/pagmo2-2.10-161-g0a589425/tests/fork_island.cpp:30:
In file included from /usr/local/include/boost/test/included/unit_test.hpp:23:
/usr/local/include/boost/test/impl/execution_monitor.ipp:560:10: error: use of undeclared identifier 'SIGIO'
    case SIGPOLL:
         ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:145:23: note: expanded from macro 'SIGPOLL'
#      define SIGPOLL SIGIO
                      ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:669:22: error: no member named 'sa_sigaction' in 'sigaction'
    if( m_new_action.sa_sigaction || m_new_action.sa_handler ) {
        ~~~~~~~~~~~~ ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:674:34: error: use of undeclared identifier 'SA_SIGINFO'
    m_new_action.sa_flags     |= SA_SIGINFO;
                                 ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:675:18: error: no member named 'sa_sigaction' in 'sigaction'
    m_new_action.sa_sigaction  = attach_dbg ? &boost_execution_monitor_attaching_signal_handler
    ~~~~~~~~~~~~ ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:681:34: error: use of undeclared identifier 'SA_ONSTACK'
        m_new_action.sa_flags |= SA_ONSTACK;
                                 ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:759:18: error: use of undeclared identifier 'SIGIO'
, m_POLL_action( SIGPOLL, catch_system_errors,      attach_dbg, alt_stack )
                 ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:145:23: note: expanded from macro 'SIGPOLL'
#      define SIGPOLL SIGIO
                      ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:773:9: error: unknown type name 'stack_t'; did you mean '__stack_t'?
        stack_t sigstk;
        ^
/usr/include/sys/signal.h:437:8: note: '__stack_t' declared here
struct __stack_t {
       ^
In file included from /usr/ports/devel/pagmo2/work/pagmo2-2.10-161-g0a589425/tests/fork_island.cpp:30:
In file included from /usr/local/include/boost/test/included/unit_test.hpp:23:
/usr/local/include/boost/test/impl/execution_monitor.ipp:774:41: error: unknown type name 'stack_t'; did you mean '__stack_t'?
        std::memset( &sigstk, 0, sizeof(stack_t) );
                                        ^
/usr/include/sys/signal.h:437:8: note: '__stack_t' declared here
struct __stack_t {
       ^
In file included from /usr/ports/devel/pagmo2/work/pagmo2-2.10-161-g0a589425/tests/fork_island.cpp:30:
In file included from /usr/local/include/boost/test/included/unit_test.hpp:23:
/usr/local/include/boost/test/impl/execution_monitor.ipp:776:34: error: no member named 'sigaltstack' in the global namespace
        BOOST_TEST_SYS_ASSERT( ::sigaltstack( 0, &sigstk ) != -1 );
                               ~~^
/usr/local/include/boost/test/execution_monitor.hpp:498:59: note: expanded from macro 'BOOST_TEST_SYS_ASSERT'
#define BOOST_TEST_SYS_ASSERT( cond ) BOOST_TEST_I_ASSRT( cond, ::boost::system_error( BOOST_STRINGIZE( exp ) ) )
                                                          ^~~~
/usr/local/include/boost/test/detail/throw_exception.hpp:60:44: note: expanded from macro 'BOOST_TEST_I_ASSRT'
#define BOOST_TEST_I_ASSRT( cond, ex ) if( cond ) {} else BOOST_TEST_I_THROW( ex )
                                           ^~~~
In file included from /usr/ports/devel/pagmo2/work/pagmo2-2.10-161-g0a589425/tests/fork_island.cpp:30:
In file included from /usr/local/include/boost/test/included/unit_test.hpp:23:
/usr/local/include/boost/test/impl/execution_monitor.ipp:778:31: error: use of undeclared identifier 'SS_DISABLE'
        if( sigstk.ss_flags & SS_DISABLE ) {
                              ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:780:31: error: use of undeclared identifier 'SIGSTKSZ'
            sigstk.ss_size  = BOOST_TEST_ALT_STACK_SIZE;
                              ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:179:39: note: expanded from macro 'BOOST_TEST_ALT_STACK_SIZE'
#    define BOOST_TEST_ALT_STACK_SIZE SIGSTKSZ
                                      ^
/usr/local/include/boost/test/impl/execution_monitor.ipp:782:38: error: no member named 'sigaltstack' in the global namespace
            BOOST_TEST_SYS_ASSERT( ::sigaltstack( &sigstk, 0 ) != -1 );
                                   ~~^
/usr/local/include/boost/test/execution_monitor.hpp:498:59: note: expanded from macro 'BOOST_TEST_SYS_ASSERT'
#define BOOST_TEST_SYS_ASSERT( cond ) BOOST_TEST_I_ASSRT( cond, ::boost::system_error( BOOST_STRINGIZE( exp ) ) )
                                                          ^~~~
/usr/local/include/boost/test/detail/throw_exception.hpp:60:44: note: expanded from macro 'BOOST_TEST_I_ASSRT'
#define BOOST_TEST_I_ASSRT( cond, ex ) if( cond ) {} else BOOST_TEST_I_THROW( ex )
                                           ^~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

**Comment by darioizzo** [comment]
We will need more info if you want us to help. Note that pagmo compiles on different platform as our CI can show. So this is probably an issue on the compiler / system  interaction.

Are you trying to compile the tests? Or your own code? What system and compiler are you on?

**Comment by yurivict** [comment]
System: FreeBSD 11.2 amd64
Compiler: clang-6
Compiling tests using ```-DPAGMO_BUILD_TESTS:BOOL=ON```.

**Comment by bluescarni** [comment]
This looks related:

https://sigrok.org/bugzilla/show_bug.cgi?id=649

It seems like the issue might be that we are defining the ``_POSIX_C_SOURCE`` macro when compiling some tests, and this ends up triggering some issue in the C++ library implementation.

Would you mind pasting a longer error message containing also the command line invocation of the compiler?

**Comment by yurivict** [comment]
Here is the log: https://people.freebsd.org/~yuri/pagmo2-test-2.log

**Comment by bluescarni** [comment]
Thanks for posting the log, that is a pretty strong indication that the ``_POSIX_C_SOURCE`` definition is the culprit.

The good news is that we already stopped defining that macro in the development version of pagmo2, the bad news is that it will take probably a couple of weeks before we have a new release available.

Could you test the latest git head after I have merged the relevant changes? I can ping you here when that happens.

**Comment by yurivict** [comment]
I used ```2.10-161-g0a589425```.

**Comment by bluescarni** [comment]
@yurivict The new pagmo release, 2.11, is out. Could you please re-try a FreeBSD build?

**Comment by bluescarni** [comment]
Will close this for now, as it should be fixed in pagmo>=2.11. Please re-open if needed.
